### PR TITLE
Unit 4 - CLIP and relatives - Models - OWL-ViT & CLIP

### PR DESCRIPTION
**Pull request by mattmdjaga**
This section is part of showcasing of CLIP relatives and it covers CLIP and OWL-ViT, an open vocabulary object detection model. The aim is to briefly introduce the model, show how to use it and provide resources to find more about the models. 

The whole section structure:
Models:
- CLIP (Image & text embedding) @pedro
- Donut or Nougat (Document Analysis) 
- GroupViT or OneFormer (Segmentation) 
- BLIP (Vision language with text generation) 
- OWL-VIT (Vision Language object detection)

### CODE REVIEW

**Comment by pedrogengo** [comment]
@mattmdjaga  can you change the PR title to reflect we are putting CLIP and OWL here?

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Contrastive Language-Image Pre-training (CLIP)
+
+## Introduction
+
+CLIP is a neural network adept at grasping visual concepts through natural language supervision. It operates by concurrently training a text encoder and an image feature extractor, focusing on a pretraining task that involves matching captions with corresponding images. This architecture allows CLIP to seamlessly adapt to a variety of visual classification benchmarks. It does so by simply receiving the names of the visual categories to be recognized, demonstrating "zero-shot" learning capabilities akin to those observed in GPT-2 and GPT-3 models.
+
+## Contrastive pre-training
+
+Given a batch of (image, text) pairs, CLIP computes the dense cosine similarity matrix between all possible (image, text) candidates within this batch. The core idea is too maximize the similarity between the correct pairs (showed in blue in the figure below) and minimize the similarity for incorrect pairs (showed in grey in the image). To do it, they optimize a symmetric cross entropy loss over these similarity scores.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:9]:
```suggestion
Given a batch of (image, text) pairs, CLIP computes the dense cosine similarity matrix between all possible (image, text) candidates within this batch. The core idea is to maximize the similarity between the correct pairs (shown in blue in the figure below) and minimize the similarity for incorrect pairs (shown in grey in the image). To do it, they optimize a symmetric cross-entropy loss over these similarity scores. 
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+## Contrastive pre-training
+
+Given a batch of (image, text) pairs, CLIP computes the dense cosine similarity matrix between all possible (image, text) candidates within this batch. The core idea is too maximize the similarity between the correct pairs (showed in blue in the figure below) and minimize the similarity for incorrect pairs (showed in grey in the image). To do it, they optimize a symmetric cross entropy loss over these similarity scores. 
+
+![CLIP contrastive pre-training](https://images.openai.com/blob/fbc4f633-9ad4-4dc2-bd94-0b6f1feee22f/overview-a.svg)
+
+Explaining in simple terms, we want to make the similarity between the image and its corresponding caption higher as we can, while the similarity between the image and the others captions should be small. We apply this logic to the caption too, so we want to maximize the similarity of the caption with its corresponding image, and minimize between all other images.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:13]:
```suggestion
Explaining in simple terms, we want to make the similarity between the image and its corresponding caption as high as we can, while the similarity between the image and the other captions should be small. We apply this logic to the caption too, so we want to maximize the similarity of the caption with its corresponding image and minimize between all other images.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+## Text encoder and Image encoder
+
+CLIP's design features independent encoders for images and text, allowing flexibility in their choice. Users can switch the standard image encoder, like a Vision Transformer, for alternatives like ResNet, or opt for different text processing models, enhancing adaptability and experimentation. Of course, if you switch one of the encoder, you will need to train your model again, as your embeddings distribution will be different.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:17]:
```suggestion
CLIP's design features independent encoders for images and text, allowing flexibility in their choice. Users can switch the standard image encoder, like a Vision Transformer, for alternatives like ResNet, or opt for different text processing models, enhancing adaptability and experimentation. Of course, if you switch one of the encoders, you will need to train your model again, as your embeddings distribution will be different.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+## Use cases
+CLIP, can be leveraged for a variety of applications. Here are some notable use cases:
+
+- Zero shot image classification;
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:22]:
```suggestion
- Zero-shot image classification;
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+image = Image.open(requests.get(url, stream=True).raw)
+
+inputs = processor(text=["a photo of a cat", "a photo of a dog"], images=image, return_tensors="pt", padding=True)
+
+outputs = model(**inputs)
+logits_per_image = outputs.logits_per_image 
+probs = logits_per_image.softmax(dim=1)
+```
+
+After executing this code, we got the following probabilites:
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:51]:
```suggestion
After executing this code, we got the following probabilities:
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+## Text encoder and Image encoder
+
+CLIP's design features independent encoders for images and text, allowing flexibility in their choice. Users can switch the standard image encoder, like a Vision Transformer, for alternatives like ResNet, or opt for different text processing models, enhancing adaptability and experimentation. Of course, if you switch one of the encoder, you will need to train your model again, as your embeddings distribution will be different.
+
+## Use cases
+CLIP, can be leveraged for a variety of applications. Here are some notable use cases:
+
+- Zero shot image classification;
+- Similarity search;
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:23]:
```suggestion
- Similarity search.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+Despite CLIP's proficiency in zero-shot classification, it is unlikely to outperform a specialized, fine-tuned model. Moreover, its generalization capabilities are somewhat limited, particularly in scenarios involving data or examples not encountered during training.
+Th paper also shows that CLIP's effectiveness and biases are impacted by the choice of categories, as demonstrated in tests using the Fairface dataset. Notable disparities were found in gender and racial classifications, with gender accuracy over 96% and racial accuracy around 93%.
+
+## Conclusion
+
+In conclusion, the CLIP model from OpenAI has revolutionized the multimodal field. What sets CLIP apart is its proficiency in zero-shot learning, allowing it to classify images into categories it wasn't explicitly trained on. This remarkable ability to generalize comes from its innovative training method, where it learns to match images with text captions.
+
+## References
+
+- https://arxiv.org/pdf/2103.00020.pdf
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:66]:
```suggestion
- [CLIP paper](https://arxiv.org/abs/2103.00020)
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+- https://lilianweng.github.io/posts/2021-05-31-contrastive/#clip
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:67]:
```suggestion
- [Lilian Weng's blog post](https://lilianweng.github.io/posts/2021-05-31-contrastive/#clip)
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+image = Image.open(requests.get(url, stream=True).raw)
+
+inputs = processor(text=["a photo of a cat", "a photo of a dog"], images=image, return_tensors="pt", padding=True)
+
+outputs = model(**inputs)
+logits_per_image = outputs.logits_per_image 
+probs = logits_per_image.softmax(dim=1)
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:48]: I think it would be nice to add (for example) an illustration showing the image and the probs from this code.

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/owl_vit.mdx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	frameborder="0"
+	width="850"
+	height="450">
+</iframe>
+
+### Conclusion
+
+OWL-ViT's approach to object detection represents a notable shift in how AI models understand and interact with the visual world. By integrating language understanding with visual perception, it pushes the boundaries of object detection, enabling more accurate and versatile models capable of identifying a broader range of objects. This evolution in model capabilities is crucial for applications requiring nuanced understanding of visual scenes, particularly in dynamic, real-world environments.
+
+For more in-depth information and technical details, refer to the [OWL-VIT paper](https://arxiv.org/abs/2205.06230). You can also find information about [OWL-ViT 2 model](https://huggingface.co/docs/transformers/model_doc/owlv2) in the Hugging face documentation.
```
Comment (mmhamdy) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/owl_vit.mdx:75]: Nice! I was just going to ask about OWL-ViT 2 😅

**Review by mmhamdy — COMMENTED** [review]
Looking good! you briefly introduced the two models nicely. 
Side question 😅: For zero-shot object detection, what models other than OWL-ViT do you think are popular?

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx (hunk 8 above)
Comment (ricard-inho) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:67]:
```suggestion
- [Learning Transferable Visual Models From Natural Language Supervision](https://arxiv.org/pdf/2103.00020.pdf)
- [CLIP by Lilian Weng](https://lilianweng.github.io/posts/2021-05-31-contrastive/#clip)
```

**Review by ricard-inho — COMMENTED** [review]
Looks good 👍. Just added some small suggestions

**Comment by mattmdjaga** [comment]
I added a link to my YOLO mention. However I linked https://github.com/ultralytics/ultralytics instead of the YOLO paper as I think it's more relevant to the topic. Not sure if everyone else agrees with that?

**Comment by merveenoyan** [comment]
@mattmdjaga you can add YOLO paper and add ultralytics in references. If you can commit rest of my suggestions I'll give another review and we can merge 😊

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+- Similarity search;
+- Diffusion models conditioning.
+
+## Usage
+
+For practical applications, one typically uses an image, and pre-defined classes as input. The provided Python example demonstrates how to use the transformers library for running CLIP. In this example, we want to zero-shot classify the image below between `dog` or `cat`.
+
+![A photo of cats](http://images.cocodataset.org/val2017/000000039769.jpg)
+
+```python
```
[the remark this answers is not in the document]
Comment (pedrogengo) [reply chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:32]: I don't know if I get the idea here. Can you point me the DETR section?

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 12]
```diff
@@ -0,0 +1,67 @@
+# Contrastive Language-Image Pre-training (CLIP)
+
+## Introduction
+
+CLIP is a neural network adept at grasping visual concepts through natural language supervision. It operates by concurrently training a text encoder and an image encoder, focusing on a pretraining task that involves matching captions with corresponding images. This architecture allows CLIP to adapt to a variety of visual classification benchmarks seamlessly. It does so by simply receiving the names of the visual categories to be recognized, demonstrating "zero-shot" learning capabilities akin to those observed in GPT-2 and GPT-3 models.
```
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:5]:
```suggestion
CLIP is a neural network adept at grasping visual concepts through natural language supervision. It operates by concurrently training a text encoder and an image encoder, focusing on a pretraining task that involves matching captions with corresponding images. This architecture allows CLIP to seamlessly adapt to various visual classification benchmarks. It does so by simply receiving the names of the visual categories to be recognized, demonstrating "zero-shot" learning capabilities akin to those observed in GPT-2 and GPT-3 models.
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+![CLIP contrastive pre-training](https://images.openai.com/blob/fbc4f633-9ad4-4dc2-bd94-0b6f1feee22f/overview-a.svg)
+
+Explaining in simple terms, we want to make the similarity between the image and its corresponding caption as high as we can, while the similarity between the image and the other captions should be small. We apply this logic to the caption too, so we want to maximize the similarity of the caption with its corresponding image, and minimize between all other images.
+
+## Text Encoder and Image Encoder
+
+CLIP's design features independent encoders for images and text, allowing flexibility in their choice. Users can switch the standard image encoder, like a Vision Transformer, for alternatives like ResNet, or opt for different text encoders, enhancing adaptability and experimentation. Of course, if you switch one of the encoders, you will need to train your model again, as your embedding distribution will be different.
+
+## Use cases
```
Comment (merveenoyan) [inline chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:19]:
```suggestion
## Use Cases
```

File: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx (hunk 11 above)
[the remark this answers is not in the document]
Comment (merveenoyan) [reply chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/clip.mdx:32]: I meant you could add a small implementation of CLIP architecture in PyTorch, nothing extensive

**Review by merveenoyan — APPROVED** [review]
I think we can get this merged @mattmdjaga and if you feel like it you can add a small PyTorch implementation of CLIP in another PR or this PR, what do you think?

**Comment by mattmdjaga** [comment]
> I think we can get this merged @mattmdjaga and if you feel like it you can add a small PyTorch implementation of CLIP in another PR or this PR, what do you think?

Yeh I'm down to add a PyTroch CLIP implementation and it can be in another PR as I might first need to work on other parts of this chapter which haven't been completed yet.

**Comment by merveenoyan** [comment]
@mattmdjaga sure, no worries! this one is good as is

**Review by notvicent3 — APPROVED** [review]

**Review by mmhamdy — APPROVED** [review]
